text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+| x86 | `reg` | None | `i32` | `eax` |
+| x86 | `reg` | None | `i64` | `rax` |
+| x86-32 | `reg_abcd` | `b` | Any | `al` |
+| x86-64 | `reg` | `b` | Any | `al` |
+| x86 | `reg_abcd` | `h` | Any | `ah` |
+| x86 | `reg` | `w` | Any | `ax` |
+| x86 | `reg` | `k` | Any | `eax` |
+| x86-64 | `reg` | `q` | Any | `rax` |
+| x86 | `vreg` | None | `i32`, `i64`, `f32`, `f64`, `v128` | `xmm0` |
+| x86 (AVX) | `vreg` | None | `v256` | `ymm0` |
+| x86 (AVX-512) | `vreg` | None | `v512` | `zmm0` |
+| x86 (AVX-512) | `kreg` | None | Any | `k1` |
+| AArch64 | `reg` | None | Any | `x0` |
+| AArch64 | `reg` | `w` | Any | `w0` |
+| AArch64 | `reg` | `x` | Any | `x0` |
+| AArch64 | `vreg` | None | Any | `v0` |
+| AArch64 | `vreg` | `b` | Any | `b0` |
+| AArch64 | `vreg` | `h` | Any | `h0` |
+| AArch64 | `vreg` | `s` | Any | `s0` |
+| AArch64 | `vreg` | `d` | Any | `d0` |
+| AArch64 | `vreg` | `q` | Any | `q0` |
+| ARM | `reg` | None | Any | `r0` |
+| ARM | `vreg` | None | `f32` | `s0` |
+| ARM | `vreg` | None | `f64`, `v64` | `d0` |
+| ARM | `vreg` | None | `v128` | `q0` |
+| ARM | `vreg` | `e` / `f` | `v128` | `d0` / `d1` |
+| RISC-V | `reg` | None | Any | `x1` |
+| RISC-V | `vreg` | None | Any | `f0` |
+
+> Notes:
+> - on ARM `e` / `f`: this prints the low or high doubleword register name of a NEON quad (128-bit) register.
+> - on AArch64 `reg`: a warning is emitted if the input type is smaller than 64 bits, suggesting to use the `w` modifier. The warning can be suppressed by explicitly using the `x` modifier.
+
+[llvm-argmod]: http://llvm.org/docs/LangRef.html#asm-template-argument-modifiers
+
+## Flags
+
+Flags are used to further influence the behavior of the inline assembly block.
+Currently the following flags are defined:
+- `pure`: The `asm` block has no side effects, and its outputs depend only on its direct inputs (i.e. the values themselves, not what they point to). This allows the compiler to execute the `asm` block fewer times than specified in the program (e.g. by hoisting it out of a loop) or even eliminate it entirely if the outputs are not used. A warning is emitted if this flag is used on an `asm` with no outputs.
+- `nomem`: The `asm` blocks does not read or write to any memory. This allows the compiler to cache the values of modified global variables in registers across the `asm` block since it knows that they are not read or written to by the `asm`.
+- `readonly`: The `asm` block does not write to any memory. This allows the compiler to cache the values of unmodified global variables in registers across the `asm` block since it knows that they are not written to by the `asm`.
+- `preserves_flags`: The `asm` block does not modify the flags register (defined below). This allows the compiler to avoid recomputing the condition flags after the `asm` block.
+- `noreturn`: The `asm` block never returns, and its return type is defined as `!` (never). Behavior is undefined if execution falls through past the end of the asm code.
```

Comment:
> its return type is defined as `!` (never)

Ordinarily does `asm!` have a return type of `()`?

Reply:
Also, is there any difference between using `noreturn` and putting a call to [`unreachable_unchecked`](https://doc.rust-lang.org/std/hint/fn.unreachable_unchecked.html) after the `asm!` block?

Reply:
Yes, `asm!` has a return type of `()`.

> Also, is there any difference between using noreturn and putting a call to unreachable_unchecked after the asm! block?

If you look at the "Mapping to LLVM IR" section, you'll see that's exactly what it does.

Reply:
Yes, I saw this:

> If the `noreturn` flag is set then an `unreachable` LLVM instruction is inserted after the asm invocation.

But I think `unreachable_unchecked` will result in the same code (modulo inlining, etc.). Do you think this has a performance impact, or is there another reason other than brevity to prefer the `noreturn` modifier?

Reply:
I added the `noreturn` flag because I feel that it makes it more explicit that the asm never returns.

Reply:
There’s another reason to want this here: naked functions. Those are only allowed to have a single asm statement - anything else is UB. This means that currently, you can’t have a naked function that returns !, which is pretty inconvenient.

With this noreturn flag, the problem is neatly solved.

Reply:
Umm... can't you just stick a panic at the end of the function to make it return never?

Reply:
Naked functions can't have a panic at the end of the function. Again, naked functions are **only** allowed to have a single asm statement - nothing else.

Reply:
What is the advantage of a naked function over a `global_asm!` block?

Reply:
The main advantage is that the compiler handles name mangling for you.

Reply:
IMO, the main advantage is that it is inline, which is much easier to read than jumping to another file and makes it easier to deal with things like accessing globals or modifying rust variables.

@roblabla Is this documented somewhere? I've used a lot of naked functions with function calls etc, and it seems to work fine as long as your asm sets up the stack properly.

Reply:
Btw, I'm glad to discuss naked asm further, but perhaps we should do it in the tracking issue instead?

Reply:
> IMO, the main advantage is that it is inline, which is much easier to read than jumping to another file

You don't have to place `global_asm!();` in a different file.

>  and makes it easier to deal with things like accessing globals or modifying rust variables.

:+1:

Reply:
@mark-i-m 

>it is generally unsafe to write anything but inline assembly inside a naked function.  The [LLVM language reference](http://llvm.org/docs/LangRef.html#function-attributes) describes this feature as having "very system-specific consequences", which the programmer must
be aware of.

https://github.com/rust-lang/rfcs/pull/1201/files#diff-e089b5aa9b57392ee8a5efa292856d1eR74

There is [an open RFC](https://github.com/rust-lang/rfcs/pull/2774/files) to amend 1201 to explicitly deny various common misuses of it. Let's move naked function discussion there.